plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,249 @@
+        this.writeBufferSize = exchangeStorage.getWriteBufferSizeInBytes();
+
+        for (int i = 0; i < numBuffers; ++i) {
+            freeBuffers.add(Slices.allocate(writeBufferSize).getOutput());
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (freeBuffers.isEmpty() && pendingBuffers.size() < outputPartitionCount) {
+            synchronized (this) {
+                if (blockedFuture.isDone()) {
+                    blockedFuture = new CompletableFuture<>();
+                }
+                return blockedFuture;
+            }
+        }
+        else {
+            return NOT_BLOCKED;
+        }
+    }
+
+    @Override
+    public void add(int partitionId, Slice data)
+    {
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!committed, "already committed");
+        if (closed) {
+            return;
+        }
+
+        writers.computeIfAbsent(partitionId, this::createWriter);
+        synchronized (writers.get(partitionId)) {
+            writeToExchangeStorage(partitionId, Slices.wrappedIntArray(data.length()));
+            writeToExchangeStorage(partitionId, data);
+        }
+    }
+
+    private ExchangeStorageWriter createWriter(int partitionId)
+    {
+        URI outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return exchangeStorage.createExchangeStorageWriter(outputPath, secretKey);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeToExchangeStorage(int partitionId, Slice slice)
+    {
+        int position = 0;
+        while (position < slice.length()) {
+            SliceOutput pendingBuffer = pendingBuffers.computeIfAbsent(partitionId, ignored -> {
+                try {
+                    return freeBuffers.take();
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            });
+            int writableBytes = min(pendingBuffer.writableBytes(), slice.length() - position);
+            pendingBuffer.writeBytes(slice.getBytes(position, writableBytes));
+            position += writableBytes;
+
+            flushIfNeeded(partitionId, false);
+        }
+    }
+
+    private void flushIfNeeded(int partitionId, boolean finished)
+    {
+        SliceOutput buffer = pendingBuffers.get(partitionId);
+        if (!buffer.isWritable() || finished) {
+            if (!buffer.isWritable()) {
+                pendingBuffers.remove(partitionId);
+            }
+            try {
+                writers.get(partitionId).write(buffer.slice()).addListener(() -> {
+                    buffer.reset();
+                    freeBuffers.add(buffer);
```

Comment:
The state of the `freeBuffers` queue and the `blockedFuture` inherently coupled. It feels like mutation of these two fields has to be done under the same lock

Reply:
ditto. Will try to use a global lock.

Reply:
I'm not sure locking and BlockingQueue can be used together. Doesn't this cause deadlocks?

Reply:
Actually yeah, you are right. It may cause a deadlock